File: sdk/typescript/src/local/CrlCache.test.ts
import { describe, it, expect, vi } from 'vitest';
import { CrlCache, type CrlData } from './CrlCache';

// A controllable clock so TTL behavior is deterministic (no real timers).
function clock(start = 0) {
  let t = start;
  return { now: () => t, advance: (ms: number) => (t += ms) };
}

const LIST_A: CrlData = { entries: [{ agentId: 'agent-1', reason: 'key compromise' }] };
const LIST_B: CrlData = { entries: [{ agentId: 'agent-2' }] };

describe('CrlCache', () => {
  it('serves a fetched list while fresh and null once stale', async () => {
    const c = clock();
    const cache = new CrlCache({ fetch: async () => LIST_A, ttlMs: 1000, now: c.now });

    expect(cache.current()).toBeNull(); // nothing fetched yet
    expect(cache.status().loaded).toBe(false);

    await cache.refreshNow();
    expect(cache.current()).toEqual(LIST_A);
    expect(cache.isFresh()).toBe(true);

    c.advance(1000); // exactly at TTL is still fresh (age <= ttl)
    expect(cache.current()).toEqual(LIST_A);

    c.advance(1); // now beyond TTL
    expect(cache.current()).toBeNull();
    expect(cache.isFresh()).toBe(false);
    expect(cache.status().loaded).toBe(true); // still loaded, just stale
  });

  it('retains the previous list and records the error when a refresh fails', async () => {
    const c = clock();
    let call = 0;
    const cache = new CrlCache({
      fetch: async () => {
        call += 1;
        if (call === 1) return LIST_A;
        throw new Error('feed down');
      },
      ttlMs: 1000,
      now: c.now,
      onError: () => {}, // silence in test
    });

    await cache.refreshNow();
    expect(cache.current()).toEqual(LIST_A);

    c.advance(500);
    await cache.refreshNow(); // fails
    expect(cache.current()).toEqual(LIST_A); // previous list retained within TTL
    expect(cache.status().lastError).toBeInstanceOf(Error);

    c.advance(501); // past TTL from the original successful fetch
    expect(cache.current()).toBeNull(); // stale -> caller applies onStale
  });

  it('updates to the newest list on a successful refresh', async () => {
    const c = clock();
    let list = LIST_A;
    const cache = new CrlCache({ fetch: async () => list, ttlMs: 1000, now: c.now });
    await cache.refreshNow();
    expect(cache.current()).toEqual(LIST_A);
    list = LIST_B;
    await cache.refreshNow();
    expect(cache.current()).toEqual(LIST_B);
  });

  it('coalesces concurrent refreshes into a single fetch', async () => {
    const fetch = vi.fn(async () => {
      await new Promise((r) => setTimeout(r, 5));
      return LIST_A;
    });
    const cache = new CrlCache({ fetch, ttlMs: 1000 });
    await Promise.all([cache.refreshNow(), cache.refreshNow(), cache.refreshNow()]);
    expect(fetch).toHaveBeenCalledTimes(1);
  });

  it('rejects an invalid fetch shape without throwing on refresh', async () => {
    const cache = new CrlCache({
      // @ts-expect-error intentionally wrong shape
      fetch: async () => ({ nope: true }),
      ttlMs: 1000,
      onError: () => {},
    });
    await expect(cache.refreshNow()).resolves.toBeUndefined();
    expect(cache.current()).toBeNull();
    expect(cache.status().lastError).toBeInstanceOf(TypeError);
  });

  it('validates configuration', () => {
    // @ts-expect-error missing fetch
    expect(() => new CrlCache({})).toThrow(TypeError);
    expect(() => new CrlCache({ fetch: async () => LIST_A, ttlMs: 0 })).toThrow(RangeError);
    expect(
      () => new CrlCache({ fetch: async () => LIST_A, ttlMs: 1000, refreshIntervalMs: 0 }),
    ).toThrow(RangeError);
  });

  it('defaults onStale to soft-open and honors an explicit reject policy', () => {
    expect(new CrlCache({ fetch: async () => LIST_A }).onStale).toBe('soft-open');
    expect(new CrlCache({ fetch: async () => LIST_A, onStale: 'reject' }).onStale).toBe('reject');
  });

  it('start() refreshes immediately and stop() is idempotent', async () => {
    const fetch = vi.fn(async () => LIST_A);
    const cache = new CrlCache({ fetch, ttlMs: 10_000, refreshIntervalMs: 10_000 });
    cache.start();
    await Promise.resolve(); // let the immediate refresh settle
    await cache.refreshNow();
    expect(fetch).toHaveBeenCalled();
    cache.stop();
    cache.stop(); // no throw
  });
});

File: sdk/java/src/test/java/org/opena2a/aim/atx/CrlCacheTest.java
package org.opena2a.aim.atx;

import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class CrlCacheTest {

    private static AtxTrustAnchors.Crl list(String agentId) {
        return new AtxTrustAnchors.Crl(List.of(new AtxTrustAnchors.Crl.Entry(agentId, "key compromise")));
    }

    @Test
    void servesWhileFreshAndNullOnceStale() {
        AtomicLong clock = new AtomicLong(0);
        CrlCache cache = CrlCache.builder(() -> list("agent-1"))
                .ttl(Duration.ofMillis(1000))
                .nowMillis(clock::get)
                .build();

        assertNull(cache.current(), "nothing fetched yet");
        assertFalse(cache.status().loaded());

        cache.refreshNow();
        assertNotNull(cache.current());
        assertTrue(cache.isFresh());

        clock.set(1000); // exactly at TTL is still fresh (age <= ttl)
        assertNotNull(cache.current());

        clock.set(1001); // beyond TTL
        assertNull(cache.current());
        assertFalse(cache.isFresh());
        assertTrue(cache.status().loaded(), "still loaded, just stale");
    }

    @Test
    void retainsPreviousListAndRecordsErrorWhenRefreshFails() {
        AtomicLong clock = new AtomicLong(0);
        AtomicLong call = new AtomicLong(0);
        CrlCache cache = CrlCache.builder(() -> {
                    if (call.incrementAndGet() == 1) {
                        return list("agent-1");
                    }
                    throw new RuntimeException("feed down");
                })
                .ttl(Duration.ofMillis(1000))
                .nowMillis(clock::get)
                .onError(t -> { /* silence in test */ })
                .build();

        cache.refreshNow();
        assertNotNull(cache.current());

        clock.set(500);
        cache.refreshNow(); // fails
        assertNotNull(cache.current(), "previous list retained within TTL");
        assertNotNull(cache.status().lastError());

        clock.set(1001); // past TTL from the original successful fetch
        assertNull(cache.current(), "stale -> caller applies stale policy");
    }

    @Test
    void updatesToNewestListOnSuccessfulRefresh() {
        AtomicLong clock = new AtomicLong(0);
        AtomicReference<AtxTrustAnchors.Crl> source = new AtomicReference<>(list("agent-1"));
        CrlCache cache = CrlCache.builder(source::get)
                .ttl(Duration.ofMillis(1000))
                .nowMillis(clock::get)
                .build();

        cache.refreshNow();
        assertEquals("agent-1", cache.current().entries().get(0).agentId());

        source.set(list("agent-2"));
        cache.refreshNow();
        assertEquals("agent-2", cache.current().entries().get(0).agentId());
    }

    @Test
    void nullFromSourceIsCapturedNotThrown() {
        CrlCache cache = CrlCache.builder(() -> null)
                .ttl(Duration.ofMillis(1000))
                .onError(t -> { })
                .build();
        cache.refreshNow(); // must not throw
        assertNull(cache.current());
        assertNotNull(cache.status().lastError());
    }

    @Test
    void validatesConfiguration() {
        assertThrows(NullPointerException.class, () -> CrlCache.builder(null));
        assertThrows(IllegalArgumentException.class,
                () -> CrlCache.builder(() -> list("a")).ttl(Duration.ZERO).build());
        assertThrows(IllegalArgumentException.class,
                () -> CrlCache.builder(() -> list("a")).refreshInterval(Duration.ZERO).build());
    }

    @Test
    void defaultsToSoftOpenAndHonorsReject() {
        assertEquals(CrlCache.StalePolicy.SOFT_OPEN,
                CrlCache.builder(() -> list("a")).build().onStale());
        assertEquals(CrlCache.StalePolicy.REJECT,
                CrlCache.builder(() -> list("a")).onStale(CrlCache.StalePolicy.REJECT).build().onStale());
    }

    @Test
    void startRefreshesAndStopIsIdempotent() throws Exception {
        AtomicLong calls = new AtomicLong(0);
        CrlCache cache = CrlCache.builder(() -> {
                    calls.incrementAndGet();
                    return list("agent-1");
                })
                .ttl(Duration.ofSeconds(10))
                .refreshInterval(Duration.ofSeconds(10))
                .build();
        cache.start();
        // The immediate scheduled refresh runs on a background thread; give it a moment.
        for (int i = 0; i < 50 && calls.get() == 0; i++) {
            Thread.sleep(10);
        }
        assertTrue(calls.get() >= 1, "start() should trigger an immediate refresh");
        cache.stop();
        cache.stop(); // no throw
    }

    // A Crl whose entry list is null is MALFORMED, not empty, and the cache must refuse it.
    //
    // This is the shape a decoder produces when the feed carries its list under a different
    // key — which is exactly what the AIM server does: it serves `revocations`, and this
    // record binds `entries`. Caching it marks the cache FRESH, and LocalAtxVerifier then
    // reads null entries as "nothing is revoked", so every revoked credential verifies with
    // no error, no log line, and a healthy-looking status(). That is the silent fail-open
    // the stale policy exists to prevent, arriving through a shape mismatch rather than a
    // network fault.
    @Test
    void refusesToCacheACrlWithNullEntries() {
        AtomicLong clock = new AtomicLong(0);
        AtomicReference<AtxTrustAnchors.Crl> next = new AtomicReference<>(new AtxTrustAnchors.Crl(null));

        CrlCache cache = CrlCache.builder(next::get)
                .ttl(Duration.ofMinutes(5))
                .nowMillis(clock::get)
                .build();

        cache.refreshNow();

        assertNull(cache.current(), "a malformed CRL must not become the served list");
        assertFalse(cache.isFresh(), "a malformed CRL must not mark the cache fresh");
        assertNotNull(cache.status().lastError(), "refusing a malformed CRL must surface an error");
        assertTrue(cache.status().lastError().getMessage().contains("null entries"));
    }

    // The control, and the line the guard must not cross: an EMPTY list is a legitimate
    // answer meaning "nothing is revoked right now", and it must still be cached and served.
    @Test
    void stillAcceptsAGenuinelyEmptyCrl() {
        AtomicLong clock = new AtomicLong(0);

        CrlCache cache = CrlCache.builder(() -> new AtxTrustAnchors.Crl(List.of()))
                .ttl(Duration.ofMinutes(5))
                .nowMillis(clock::get)
                .build();

        cache.refreshNow();

        assertNotNull(cache.current(), "an empty CRL is a valid answer and must be served");
        assertTrue(cache.isFresh());
        assertNull(cache.status().lastError());
        assertEquals(0, cache.current().entries().size());
    }

    // A malformed refresh must not destroy a good list that is still within its TTL.
    @Test
    void retainsThePreviousGoodListWhenARefreshIsMalformed() {
        AtomicLong clock = new AtomicLong(0);
        AtomicReference<AtxTrustAnchors.Crl> next = new AtomicReference<>(list("agent-1"));

        CrlCache cache = CrlCache.builder(next::get)
                .ttl(Duration.ofMinutes(5))
                .nowMillis(clock::get)
                .build();

        cache.refreshNow();
        assertNotNull(cache.current());

        next.set(new AtxTrustAnchors.Crl(null));
        cache.refreshNow();

        assertNotNull(cache.current(), "a malformed refresh must not discard a still-fresh good list");
        assertEquals(1, cache.current().entries().size());
        assertNotNull(cache.status().lastError(), "the malformed refresh must still be recorded");
    }
}
